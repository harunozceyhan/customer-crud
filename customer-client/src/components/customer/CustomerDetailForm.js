import React from 'react';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import DayPickerInput from 'react-day-picker/DayPickerInput';
import 'react-day-picker/lib/style.css';

const CustomerDetailForm = ({ customerDetail, validated, handleSubmit, onChange, onBack }) =>
    <div>
        <Card>
            <Card.Header>
                <h3 >{customerDetail.customerID ? "Edit" : "Add"} Customer</h3>
            </Card.Header>
            <Card.Body>

                <Form noValidate validated={validated} onSubmit={handleSubmit} >
                    <Form.Group as={Row} controlId="firstNameValidation">
                        <Form.Label column md="4">First name</Form.Label>
                        <Col sm="8">
                            <Form.Control name="firstName" type="text" placeholder="First name..." value={customerDetail.firstName} onChange={onChange} required />
                            <Form.Control.Feedback type="invalid">Field is required</Form.Control.Feedback>
                        </Col>
                    </Form.Group>
                    <Form.Group as={Row} controlId="lastNameValidation">
                        <Form.Label column md="4">Last name</Form.Label>
                        <Col sm="8">
                            <Form.Control name="lastName" type="text" placeholder="Last name..." value={customerDetail.lastName} onChange={onChange} required />
                            <Form.Control.Feedback type="invalid">Field is required</Form.Control.Feedback>
                        </Col>
                    </Form.Group>

                    <Form.Group as={Row} controlId="birthdayValidation">
                        <Form.Label column md="4">Birthday</Form.Label>
                        <Col sm="8">
                            <DayPickerInput  name="birthday" onDayChange={onChange} selectedDay={customerDetail.birthday} value={customerDetail.birthday.toString().substring(0, 10)} />
                        </Col>
                    </Form.Group>

                    <Form.Group as={Row} controlId="genderValidation">
                        <Form.Label column md="4">Gender</Form.Label>
                        <Col sm="8" as={Row}>
                            <Form.Check type="radio" className="margin-left-12">
                                <Form.Check.Input type="radio" name="gender" label="Male" value="m" checked={customerDetail.gender === "m"} onChange={onChange} required/>
                                <Form.Check.Label>Male</Form.Check.Label>
                                <Form.Control.Feedback type="invalid">Field is required</Form.Control.Feedback>
                            </Form.Check>

                            <Form.Check type="radio" className="margin-left-12">
                                <Form.Check.Input type="radio" name="gender" label="Female" value="w" checked={customerDetail.gender === "w"} onChange={onChange} required/>
                                <Form.Check.Label>Female</Form.Check.Label>
                                <Form.Control.Feedback type="invalid">Field is required</Form.Control.Feedback>
                            </Form.Check>

                        </Col>
                    </Form.Group>

                    <Form.Group as={Row} controlId="lastContactValidation">
                        <Form.Label column md="4">Last Contact</Form.Label>
                        <Col sm="8">
                            <DayPickerInput name="lastContact" onDayChange={onChange} selectedDay={customerDetail.lastContact} value={customerDetail.lastContact.toString().substring(0, 10)} />

                        </Col>
                    </Form.Group>

                    <Form.Group as={Row} controlId="lifeTimeValidation">
                        <Form.Label column md="4">Life Time</Form.Label>
                        <Col sm="8">
                            <Form.Control name="customerLifetimeValue" type="text" placeholder="Life Time Value..." value={customerDetail.customerLifetimeValue} onChange={onChange} required />
                            <Form.Control.Feedback type="invalid">Field is required</Form.Control.Feedback>
                        </Col>
                    </Form.Group>
                    <Button type="submit" variant="success" className="margin-left-12 float-right">{customerDetail.customerID ? "Update" : "Add"} Customer</Button>
                    <Button type="button" variant="secondary" onClick={onBack} className="float-right">Return To Customer List</Button>
                </Form>
            </Card.Body>
        </Card>
    </div>

export default CustomerDetailForm;