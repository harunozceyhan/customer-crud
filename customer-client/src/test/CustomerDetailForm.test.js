import React from 'react';
import { shallow, mount } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import CustomerDetailForm from '../components/customer/CustomerDetailForm'

configure({ adapter: new Adapter() });

let mockData = {
    customerID: 1,
    firstName: "Test Name",
    lastName: "Test Last name",
    gender: "w",
    customerLifetimeValue: 4.5,
    birthday: "1996-10-12",
    lastContact: "2017-06-01T23:28:56.782Z"
};

describe('CustomerDetailForm component unit test', () => {
    it('renders without crashing', () => {
        shallow(<CustomerDetailForm customerDetail={mockData} />);
    });
});

describe('CustomerDetailForm snapshot test', () => {
    it('matches the snapshot', () => {
        // shallow doesn't create component tree.
        // Instead of shallow, renderer is used to create component tree
        const tree = renderer.create(<CustomerDetailForm customerDetail={mockData} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

// Check all form element values for given mockData
describe('CustomerDetailForm Component Form Element Tests', () => {
    it('Element with name firstName value should be "Test Name"', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('input[name="firstName"]').props().value).toEqual('Test Name');
    });

    it('Element with name lastName value should be "Test Last name"', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('input[name="lastName"]').props().value).toEqual('Test Last name');
    });

    it('Birthday Element value should be "1996-10-12"', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('.DayPickerInput').first().find('input').props().value).toEqual('1996-10-12');
    });

    it('Radio Element with value "w" value should be checked', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('input[value="w"]').props().checked).toEqual(true);
    });

    it('Radio Element with value "m" value should not be checked', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('input[value="m"]').props().checked).toEqual(false);
    });


    it('LastContact Element value should be "2017-06-01"', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('.DayPickerInput').at(1).find('input').props().value).toEqual('2017-06-01');
    });

    it('Element with name lastName value should be "4.5"', () => {
        let wrapper = mount(<CustomerDetailForm customerDetail={mockData} onChange={()=>{}}/>);
        expect(wrapper.find('input[name="customerLifetimeValue"]').props().value).toEqual(4.5);
    });
});



