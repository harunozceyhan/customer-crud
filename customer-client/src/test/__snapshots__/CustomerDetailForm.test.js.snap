// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetailForm snapshot test matches the snapshot 1`] = `
<div>
  <div
    className="card"
  >
    <div
      className="card-header"
    >
      <h3>
        Edit
         Customer
      </h3>
    </div>
    <div
      className="card-body"
    >
      <form
        className=""
        noValidate={true}
      >
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            First name
          </label>
          <div
            className="col-sm-8"
          >
            <input
              className="form-control"
              id="firstNameValidation"
              name="firstName"
              placeholder="First name..."
              required={true}
              type="text"
              value="Test Name"
            />
            <div
              className="invalid-feedback"
            >
              Field is required
            </div>
          </div>
        </div>
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Last name
          </label>
          <div
            className="col-sm-8"
          >
            <input
              className="form-control"
              id="lastNameValidation"
              name="lastName"
              placeholder="Last name..."
              required={true}
              type="text"
              value="Test Last name"
            />
            <div
              className="invalid-feedback"
            >
              Field is required
            </div>
          </div>
        </div>
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Birthday
          </label>
          <div
            className="col-sm-8"
          >
            <div
              className="DayPickerInput"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                placeholder="YYYY-M-D"
                value="1996-10-12"
              />
            </div>
          </div>
        </div>
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Gender
          </label>
          <div
            className="col-sm-8 row"
          >
            <div
              className="margin-left-12 form-check"
            >
              <input
                checked={false}
                className="form-check-input"
                id="genderValidation"
                label="Male"
                name="gender"
                required={true}
                type="radio"
                value="m"
              />
              <label
                className="form-check-label"
                htmlFor="genderValidation"
                type="checkbox"
              >
                Male
              </label>
              <div
                className="invalid-feedback"
              >
                Field is required
              </div>
            </div>
            <div
              className="margin-left-12 form-check"
            >
              <input
                checked={true}
                className="form-check-input"
                id="genderValidation"
                label="Female"
                name="gender"
                required={true}
                type="radio"
                value="w"
              />
              <label
                className="form-check-label"
                htmlFor="genderValidation"
                type="checkbox"
              >
                Female
              </label>
              <div
                className="invalid-feedback"
              >
                Field is required
              </div>
            </div>
          </div>
        </div>
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Last Contact
          </label>
          <div
            className="col-sm-8"
          >
            <div
              className="DayPickerInput"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                placeholder="YYYY-M-D"
                value="2017-06-01"
              />
            </div>
          </div>
        </div>
        <div
          className="form-group row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Life Time
          </label>
          <div
            className="col-sm-8"
          >
            <input
              className="form-control"
              id="lifeTimeValidation"
              name="customerLifetimeValue"
              placeholder="Life Time Value..."
              required={true}
              type="text"
              value={4.5}
            />
            <div
              className="invalid-feedback"
            >
              Field is required
            </div>
          </div>
        </div>
        <button
          className="margin-left-12 float-right btn btn-success"
          disabled={false}
          type="submit"
        >
          Update
           Customer
        </button>
        <button
          className="float-right btn btn-secondary"
          disabled={false}
          type="button"
        >
          Return To Customer List
        </button>
      </form>
    </div>
  </div>
</div>
`;
